feat: Bubble support various message formats

gene9831 opened the pull request:
> [!IMPORTANT]  
> 此PR包含破坏性变更

- 单个气泡支持多格式消息，可以自定义渲染器来渲染不同格式消息
- 气泡支持设置形状
- 气泡列表支持设置加载中状态

TODO 
- 多格式消息如何转换成可复制的字符串
- 多格式组件消息如何使用代码主动交互

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a flexible message rendering system for bubbles, supporting custom, class-based, and component renderers.
  * Added support for multiple message formats, including text, markdown, tool status, and collapsible content.
  * Implemented a BubbleProvider component for registering and managing message renderers.
  * Enabled role-based avatars, shape customization (rounded/corner), and improved loading indicators in bubble lists.

* **Bug Fixes**
  * Improved reactivity and conditional rendering in bubble lists.

* **Documentation**
  * Expanded and clarified documentation on bubble shapes, loading states, multiple message formats, and API types.

* **Style**
  * Updated and added styles for new bubble shapes, loading indicators, and syntax highlighting.

* **Chores**
  * Added "dompurify" as a dependency for secure HTML sanitization.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gene9831 committed:
feat: add reasoning feature and shape customization to bubble component with corresponding demos and documentation

gene9831 committed:
Merge branch 'develop' of github.com:opentiny/tiny-robot into feat/bubble-inference

gene9831 committed:
fix: 更新气泡组件形状默认值为 'corner'，并修改文档以反映新的形状选项

gene9831 committed:
Merge branch 'develop' of github.com:opentiny/tiny-robot into feat/bubble-inference

gene9831 committed:
Merge branch 'develop' of github.com:opentiny/tiny-robot into feat/bubble-inference

gene9831 committed:
Merge branch 'develop' of github.com:opentiny/tiny-robot into feat/bubble-inference

gene9831 committed:
feat: implement message bubble component with various message types

- Added a new `message.vue` component to handle different message formats including text, markdown, and step-group messages.
- Removed the deprecated `reasoning.vue` component and updated documentation to reflect the new message structure.
- Enhanced the `Bubble` component to support rendering messages dynamically based on the new structure.
- Introduced a message renderer system to manage different message types effectively.

gene9831 committed:
refactor: update bubble component and remove unused SVG

- Updated the BubbleProps interface to clarify the handling of long text and words.
- Modified the BubbleTextMessageRenderer to support line breaks in message content.
- Removed the unused reasoning.svg file to clean up the assets.

gene9831 committed:
fix: improve null safety in Message component and add TODO for DOMPurify

- Updated the Message.vue component to use optional chaining for safer access to the renderer and component properties.
- Added a TODO comment in markdown.ts to address DOMPurify handling for improved security in rendering markdown content.

gene9831 committed:
refactor: update TextStep component class for improved semantics

- Changed the class name from `tr-bubble__step` to `tr-bubble__step-text` for better clarity in the TextStep component.
- This update enhances the semantic structure of the component, making it more descriptive of its purpose.

gene9831 committed:
refactor: enhance bubble message rendering and styling

- Updated the VitePress configuration to allow server access from all hosts.
- Improved the BubbleTextMessageRenderer to enhance text display with better styling properties.
- Added an IconSuccess component to the StepGroup for improved status indication.
- Adjusted padding and gap in StepGroup styles for better layout consistency.
- Enhanced ToolStep component with a new icon and improved title structure for clarity.
- Added word-break styling to TextStep for better text handling.

gene9831 committed:
refactor: enhance bubble message rendering and introduce new components

- Updated the Bubble component to utilize a content renderer for improved markdown handling.
- Introduced BubbleProvider for managing message renderers and added support for various message types.
- Replaced the deprecated message.vue with messages.vue to streamline message management.
- Added a new schema-card component for rendering schema data.
- Enhanced the markdown renderer with DOMPurify for improved security.
- Removed unused components and updated documentation to reflect the new structure.

gene9831 committed:
refactor: update Bubble component loading state and content structure

- Replaced loading spinner with an image for better visual representation.
- Streamlined the content rendering logic by simplifying template structures.
- Enhanced the handling of messages and added conditional rendering for aborted states.

gene9831 committed:
refactor: enhance message rendering and introduce custom renderers

- Updated the messages.vue component to utilize a more flexible message renderer system.
- Introduced a custom text renderer for improved message display options.
- Enhanced the markdown renderer with error handling for better robustness.
- Updated the BubbleProvider to support optional message renderers with default values.
- Improved the tool message component to handle pretty JSON formatting and added new styles for better presentation.

gene9831 committed:
refactor: streamline message content and enhance documentation

- Simplified message content in messages.vue for clarity and conciseness.
- Updated bubble.md to include detailed explanations of custom message renderers and their registration process.
- Improved documentation for better understanding of the rendering mechanism and usage examples.

gene9831 committed:
feat: enhance BubbleList component with loading state and role configuration

- Added loading state support to the BubbleList component, allowing for dynamic loading indicators based on specified roles.
- Updated loading.vue to include a loading message and a new BubbleList instance for better user feedback during data loading.
- Enhanced documentation to clarify the usage of loading and loading-role properties in the BubbleList component.

gene9831 committed:
fix: correct inheritance in custom renderer example in bubble.md

- Updated the example in bubble.md to use 'extends' instead of 'implements' for the MyRenderer class, ensuring accurate representation of class inheritance for custom message renderers.

gene9831 commented on packages/components/src/bubble/message/markdown.ts:
```diff
@@ -0,0 +1,36 @@
+    } catch (error) {
+      console.error('Error rendering markdown:', error)
+      htmlContent = options.content ?? ''
+    }
```
`DOMPurify.sanitize` used

gene9831 commented on packages/components/src/bubble/message/Message.vue:
```diff
@@ -0,0 +1,45 @@
+const props = defineProps<BubbleMessageProps>()
+const attrs = useAttrs()
+
+const getVNodeOrComponent = (type: string) => {
+  const options = { ...attrs, ...props }
+
+  const renderer = rendererMap.get(type) || fallbackRenderer
+
+  if (typeof renderer === 'function') {
+    const renderFn = renderer as BubbleMessageFunctionRenderer
+    return { isComponent: false, vNodeOrComponent: renderFn(options) }
+  }
+
+  if (renderer instanceof BubbleMessageClassRenderer) {
+    return { isComponent: false, vNodeOrComponent: renderer.render(options) }
+  }
+
+  return { isComponent: true, vNodeOrComponent: renderer }
+}
+
+const { isComponent, vNodeOrComponent } = getVNodeOrComponent(props.type)
+</script>
```
No need to reactive to type changes

gene9831 committed:
refactor: standardize class names in Bubble component for consistency

- Updated class names from `tr-bubbule` to `tr-bubble` in Bubble.vue for improved consistency and clarity in styling.
- Adjusted related CSS styles to reflect the new class naming convention.

gene9831 committed:
refactor: rename collapse message type and introduce collapsible text renderer

- Updated message type from 'collapse' to 'collapsible-text' in messages.vue for clarity.
- Removed deprecated collapse renderer and added new collapsible-text renderer for improved functionality.
- Enhanced bubble.md documentation to reflect the new renderer type and usage.
- Updated defaultRendererMap to include the new collapsible-text renderer.

gene9831 committed:
refactor: remove contentRenderer from Bubble component for simplification

- Eliminated the contentRenderer prop and related logic from Bubble.vue to streamline the rendering process.
- Updated bubble.md documentation to reflect the removal of contentRenderer and its implications for bubble content handling.
- Adjusted index.type.ts to remove references to the deprecated contentRenderer type.

hexqi commented on packages/components/src/bubble/message/tool.vue:
```diff
@@ -0,0 +1,170 @@
+  formatPretty?: boolean
+}>()
+
+const collapsed = ref(false)
```
默认可以为折叠状态collapsed true

gene9831 committed:
Revert "refactor: remove contentRenderer from Bubble component for simplification"

This reverts commit fd2176f57fd02946add0508cd0f7b6a7136ad9b9.

gene9831 committed:
refactor: update expand/collapse logic in tool component

- Renamed the `collapsed` state to `expanded` for clarity in the tool component.
- Adjusted the expand/collapse icon and conditional rendering to reflect the new state management.
- Improved readability and maintainability of the component's code.

gene9831 committed:
style: improve text wrapping in tool component and adjust slotted element selector in container

- Added `word-break: break-word;` to the tool component's CSS for better text handling.
- Modified the slotted element selector in the container component for improved flexibility in layout.

gene9831 committed:
Merge branch 'develop' of github.com:opentiny/tiny-robot into feat/bubble-inference

gene9831 committed:
feat: enhance bubble component with tool status management and new SVG icons

- Added a button to toggle tool status in the messages demo.
- Introduced reactive tool message handling with various statuses (running, success, failed, cancelled).
- Updated tool message rendering to use new SVG icons for different statuses.
- Added loading and error SVG assets, and removed the obsolete tool-running SVG.
- Included CSS animations for loading state.

gene9831 committed:
fix: enhance bubble auto-scroll functionality by adding deep watch on last bubble messages

gene9831 committed:
refactor: update bubble component to use 'content' prop instead of 'messages'

- Changed the bubble component to accept 'content' as an array of messages.
- Updated related components and documentation to reflect this change.
- Enhanced the logic for rendering messages based on the new 'content' prop.

gene9831 committed:
refactor: replace BubbleMarkdownMessageRenderer with BubbleMarkdownContentRenderer

- Updated the bubble component and related files to utilize the new BubbleMarkdownContentRenderer.
- Changed the prop names from 'messageRenderers' to 'contentRenderers' for consistency.
- Refactored the content handling to support the new structure with BubbleContentItem.
- Removed obsolete message-related files and updated documentation accordingly.
